### PR TITLE
blockdev_stream_sub_chain: Add stream test with sub snapshot chain

### PR DESCRIPTION
**Pull request by zhencliu**
data -> sn1 -> sn2 -> sn3
  block-stream: {device: sn3, base-node: sn1}

Signed-off-by: Zhenchao Liu <zhencliu@redhat.com>

ID: 1882960

### CODE REVIEW

**Comment by zhencliu** [comment]
local fs:
 (1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_stream_subchain: PASS (191.31 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-09-27T06.20-ec667ad/results.html
JOB TIME   : 196.52 s

gluster direct:
 (1/1) Host_RHEL.m8.u3.product_av.gluster_direct.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_stream_subchain: PASS (250.46 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2020-09-27T18.19-4f5ce0f/results.html
JOB TIME   : 254.17 s

iscsi direct:
 (1/1) Host_RHEL.m8.u3.product_av.iscsi_direct.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_stream_subchain: PASS (192.15 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-09-27T08.22-19706c9/results.html
JOB TIME   : 197.13 s

nbd:
 (1/1) Host_RHEL.m8.u3.product_av.nbd.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_stream_subchain: PASS (210.36 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-09-27T08.47-c4de723/results.html
JOB TIME   : 215.45 s

hi @YongxueHong @aliang123 , can you please help to review this patch? Thanks.
it deps on #2418, so please ignore the CI check error

File: qemu/tests/blockdev_stream_subchain.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        elif file_opts["driver"] == "nbd":
+            if file_params["server.port"] != file_opts["server.port"]:
+                return False
+        elif file_opts["driver"] == "rbd":
+            if file_params["image"] != file_opts["image"]:
+                return False
+        else:
+            if file_params["filename"] != file_opts["filename"]:
+                return False
+        return True
```
Comment (YongxueHong) [inline qemu/tests/blockdev_stream_subchain.py:82]:
```suggestion
        mapping = {'gluster': 'path', 'iscsi': 'lun', 
                  'nbd': 'server.port', 'rbd': 'image'}
        for driver, attr in mapping.items():
            if file_opts["driver"] == driver:
                return file_params[attr] == file_opts[attr]
            return file_params["filename"] == file_opts["filename"]
```

**Review by YongxueHong — APPROVED** [review]
LGTM

File: qemu/tests/blockdev_stream_subchain.py (hunk 1 above)
Comment (zhencliu) [reply qemu/tests/blockdev_stream_subchain.py]: yes, it's better, I made some code change and will do some tests later, thank you.

**Comment by zhencliu** [comment]
(1/1) Host_RHEL.m8.u3.product_av.iscsi_direct.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_stream_subchain: PASS (196.41 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-09-27T22.53-09dc7a7/results.html
JOB TIME   : 201.58 s
 (1/1) Host_RHEL.m8.u3.product_av.gluster_direct.luks.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.blockdev_stream_subchain.q35: PASS (246.99 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2020-09-28T11.01-186243d/results.html
JOB TIME   : 250.70 s

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.blockdev_stream_subchain.q35: PASS (315.87 s)

Acked-by: Aihua Liang <aliang@redhat.com>

**Comment by zhencliu** [comment]
hi @ngu-niny , can you please help to review this patch? Thanks.

**Comment by ngu-niny** [comment]
Test PASS on ppc:

PASS 1-Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.blockdev_stream_subchain

Acked-by: Nini Gu <ngu@redhat.com>

**Pull request merged**